### ISSUE TITLE
chest: global network limit, rework provider

### ISSUE THREAD

**Issue by bengardner** [year6b7a/item-network-factorio-mod#38]
The chest has the "provider" limit setting that prevents it from sending items to the network if there are too many already in the network.
That's great, but it can be tedious to properly set each chest with the same limit.

I'd love a global limit on the number of items that the item network can hold.
Defaults could be "unlimited".
I'd like an easy way to, say, limit iron ore to 10K or plastic to 3K without tracking down and changing a bunch of chests.

Perhaps that could be added to the network item GUI. That currently shows the number of items. 
Maybe clicking on an item can pop up a config dialog to allow setting a limit.
Maybe two numbers (current and limit) could be displayed over the icon or the format could be changed to a list.

A request chest would still try to adjust the number of items to that number by sending or taking items. That is what it currently does. (no change to requests).

A provider would hold that many items locally and send excess to the item network.
In other words, send to the item network if there are more than X items in the chest, where X defaults to 0 for each item.
The item network will refuse to accept items that would put it above the limit. (within reason)
The per-chest limit would not be used.
That's a flip of what it currently does.
I'm not sure how to tell the engine that the chest is "full" to not accept more of an item.

**Comment by year6b7a** [comment]
I like this. I'm going to play around with some implementations and see where it goes.

**Comment by Nivalux** [comment]
I agree setting global provider limits can save time during the initial setup of provider chests, as well as have it easier to adjust the limit afterwards. Otherwise you would have to adjust each chest individually.

One important thing though is, that the individual provider chest needs to retain the ability to overwrite the limit for itself.

For example: 
Set the global limit for what your main producers store in the network. 
Set the limit individually for each provider chest, that has the same item as a byproduct, so that it does not stall it's main product.

**Comment by year6b7a** [comment]
I implemented something along these lines for Network Tanks when they are configured as a Provider. They will send all contained fluids to the network up to a certain limit.

Something like that could be nice for chests, but I also want to play around with the idea of setting default buffers and limits for a specific item across all chests as a way to more flexibly manage resources.

I'm also playing around with the idea of a dynamic chest that can automatically determine buffer and limit based on usage, but that's fairly complex and hard to tune so not sure if I can figure it out.

**Comment by bengardner** [comment]
> I'm also playing around with the idea of a dynamic chest that can automatically determine buffer and limit based on usage, but that's fairly complex and hard to tune so not sure if I can figure it out.

I played with adding automatic adjustments to the request buffer.
This change in the 'take' branch of `update_network_chest` seems to work OK.
```lua
      if n_transfer > 0 then
        status = GlobalState.UPDATE_STATUS.UPDATED
        contents[request.item] = current_count + n_transfer
        GlobalState.set_item_count(request.item, network_count - n_transfer)
        -- if we transferred a full buffer and there is enough for another  
        -- in the network, then the bottleneck might be in the request buffer, 
        -- so up it by 1.
        if n_transfer == request.buffer and network_count > request.buffer then
          request.buffer = request.buffer + 1
        end
      end
```
It settled out at a little over the stack size of the inserter in some testing.

**Comment by bengardner** [comment]
I spent some time thinking about and tinkering with this.

I came up with a slight redesign that is backward compatible and opens the door for less processing.
This design is based on the availability of global limits, which I have implemented in my local copy.
I'm in the process of trying this out. So far, it looks good.

Chests can either be configured or not. They can be locked or not.
In total, there are four states:
  * unconfigured, unlocked
  * unconfigured, locked
  * configured, unlocked
  * configured, locked

## state: Unconfigured Unlocked
No requests. No filters. No bar.
On each service, the inventory is cleared and everything is sent to the network, up to the global limit.
Items in the chest are tracked (recent_items).
If the global limit is hit (the chest is not empty after being serviced), then it becomes locked.
Excess items are returned to the inventory when it becomes locked.

## state: Unconfigured Locked
No requests.
One filtered slot for each recent item that we don't have. Bar is set just above the filtered slots.
Remaining items are placed at/after the bar.
On each service, everything is sent to the network, up to the global limit.
If the chest is empty after being services, it becomes unlocked.
If the item names contained in the chest changes, then the chest is re-locked.
Re-locking pulls everything out of the chest, sets the filters, inserts the items and then sets the bar.

## state: Configured Unlocked
No filters. No bar.
On each service:
  - any items above the buffer level are sent to the network, up to the global limit. If an item fails to send, it is added to the "locked_items" table.
  - "take" requests are processes as usual
If there are any locked_items (unable to push an item to the net), then the chest becomes locked.

## state: Configured Locked
Filters are set for each "take" request, enough to hold the buffer.
Filters are set for each "give" request, enough to hold the buffer.
A single filter slot is set up for each recent item that isn't covered by the above.
The "give" requests are split. Those that can still feed the network are below the bar. Those that cannot are at/above the bar.
Same as today.
On each service, the same processing is done as the unlocked chest.
  - any items above the buffer level are sent to the network, up to the global limit. If an item fails to send, it is added to the "locked_items" table.
  - "take" requests are processes as usual
If the locked_items table is empty, then the chest becomes unlocked.
If the locked_items table changes, then the table is re-locked.
Otherwise, no change.

## Don't Reset the Chest Inventory

The key improvement is that the chest inventory is re-arranged only when transitioning states and re-locking.
That should be fairly rare.

## Leftovers
A locked chest will have at least 1 item it should, but cannot, send to the network.
I played with triggering off whether the network can accept an item instead, but that was a bit more code and less clear.
The item icon over a chest is a good indicator that it is stuck.

## Default Provider Configuration
All unconfigured items effectively have a default "give" request with buffer=0 and limit=0.
limit=0 means use the global limit.
The global limit is used to determine if the item can be sent.

## Configured Provider
Configured "providers" behave differently.
They will hold on to "buffer" items *before* sending to the network.
Setting buffer=0 is valid.
If limit=0, then the global limit is used.
If limit >0, then that limit is used, allowing a back-door to add more that the global limit to the network.
The current behavior is to send to the network until it is full and then hold onto items.

## Configured Requester
Configured "requesters" ("take") behave differently in one corner case. 
If there are more items in the chest than the buffer value, then the excess is sent to the network without checking limits.
This can be used as a hack to override the network limit. IE, set up a chest to request 1 item and then insert that item into the chest.
Otherwise, the requesters behave the same.

## Status Light
I'd like to be able to put a status light on the chest to convey the status.
I'm not sure how to do this and it is not a high priority.
I think a new entity has to be placed over the chest. Could get hairy.

Anyway, I plan on working out the kinks and posting the code.
I'm hoping this will be much more efficient.

**Comment by year6b7a** [comment]
Thanks for the detailed writeup. I'm trying to understand how this works and have a few questions.
- Is this still a Network Chest or is this a new type of chest?
- Do chests in this configuration allow for a single iron plate to be buffered and prevent more plates from being inserted?
- Is there risk that, under high volume conditions, chests shift between locked <-> unlocked in such a pattern that they have lower throughput than if the chest was configured to a fixed value?

Related, but I've been thinking of adding a High Volume Network Chest with many of the benefits you've listed here. These chests have the following properties:
- Each HVNC either requests or provides a single item. The buffer and limit values are shared globally between all chests for that item.
- The buffer rounds up to the nearest stack size so these chests don't need fancy logic for setting the bar when the buffer is full.
- Since HVNCs will update much faster than normal chests, some logic could be added to update 2 or 3 for every normal Network Chest.
- HVNCs can have a large number of slots. 100 maybe?

I'm curious if there is some overlap between these two ideas. Maybe some way to take the simplicity that comes from a single item per chest with the dynamic slot counts in your proposed chest?

**Comment by bengardner** [comment]
> Is this still a Network Chest or is this a new type of chest?

It is still a network chest.
The main play difference is that 99% of the time, I only have to worry about setting up requests (copied from assembler) and 
messing with the global limit.  I found that I don't need to click on the chest hardly ever.
The exception is that I still need to manually set requests to supply furnaces, which can't copy the recipe.
I like it.

> Do chests in this configuration allow for a single iron plate to be buffered and prevent more plates from being inserted?

Yes. Pretty much the same as what you did.

- If the global limit for iron-plates is set to 2000
- and the current network count is >= 2000
- and the chest attempts to send an iron-plate
- then the chest will become locked (if it isn't already locked)

If that chest has only contained iron-plate, then the locked chest will consist of bar=1 with the excess iron-plate items in the chest.
No filters.

If the chest also once contained, say, copper-plate, the it would set slot1 filtered to copper-plate, the bar=2 and the iron-plate beyond that.

> Is there risk that, under high volume conditions, chests shift between locked <-> unlocked in such a pattern that they have lower throughput than if the chest was configured to a fixed value?

I don't think so.
The locking prevents more items from being added to the chest. The chest is still serviced as normal.
What we gain is we aren't resetting the chest on every tick.
If we are at the edge of network capacity and toggling between locked an unlocked often, we'd still be better off.  There are generally several ticks between the state transitions.

For a complex example (mentioned above), suppose I have an unconfigured chest that is supplying both iron-plate and copper-plate. There is a furnace on each side.
When unlocked, the bar is not set and there are no filters.
If the network fills with iron-plates, then we need to stop accepting iron-plates, but keep on accepting copper-plates.
So, slot1 has a filter for copper-plates and the bar is set to 2. Excess iron-plates are at slot 2 and beyond.
That allows the chest to continue to provide copper plates.
[Side note: we could track the maximum amount of items we moved in a single tick and reserve multiple slots for an item. But I've not run into a situation that would require more than 1 slot.]
Eventually, the net can't accept any more copper-plates. At that point, the list of items in the chest changes from just iron-plate to iron-plate + copper-plate. The chest is then "relocked". Since there are no recent items that we don't have, the bar is set to 1 and excess items are beyond.
Suppose that the iron-plate limit is increased or something used iron-plates such that there is space in the net again.
The excess iron-plates are sent and the remaining inventory becomes just 'copper-plate'. Since the items changed, the chest is re-locked.  Slot 1 is filtered for iron-plate and the bar is set to 2. Copper is in slots 2+.
If we service again and send everything to the net, then the filter is cleared and the bar is reset. At that point, the chest is "unlocked" again.


As for the HVNCs, I'm not sure what the buys us, other than configuration simplicity.
Assuming we have 4 loaders pointing at a chest, all sending 360 items/sec (I remember reading that might be possible). That's 1440 items / sec that we have to dispose of. If evenly distributed, that is 24 items/tick.
Assuming iron-ore, which has a stack size of 50, we would need (1440/50) 29 slots to hold all that if we service the chest at 1 Hz.
We process 20 chests/tick @ 60 ticks per second, so, with the current code, we would hit 1 Hz with 1200 chests.

With the dual-queue patch I just sent, if the chest sends some items on each service, it would be on the "active" queue and would be serviced much more frequently.
The number of entities on that list will be short.

**Comment by bengardner** [comment]
One more comment on HVNCs.

A normal chest could be a HVNC if the buffer size for every request is a multiple of the stack size.
Then you could set the filters once and be done. Never have to change them again.
That would be neat and simplify the logic.

We could make that a check-box option.

**Comment by year6b7a** [comment]
Sounds like this addresses the two biggest issues with this mod, usability and performance. I want to figure out how to add these features to the mod in a way that's somewhat backwards compatible, easy enough for users to understand, and simple enough for me to maintain.

If there's a way to add that functionality to a normal Network Chest then that's great, but my guess is that we'll have to make a new type of chest for this functionality.

---

> I found that I don't need to click on the chest hardly ever.

This is a huge improvement over the current implementation. Exciting!

---

> But I've not run into a situation that would require more than 1 slot.

I haven't either but there definitely other factories people have built with 5k+ Network Chests that transfer huge amounts of items.

---

> A normal chest could be a HVNC if the buffer size for every request is a multiple of the stack size.

That's definitely possible but I want to avoid cluttering the existing UI too much, and an easy way to do that is to make a new type of chest.

---

I'm thinking about implementing this with 3 new chest types, a High Volume Network Requester Chest, and a High Volume Network Provider Chest, and a Network Trash Chest.

HV Requesters can be configured 99% of the time by shift-clicking from an assembler. They don't need any filter slots or bars, and they by default just pull items from the network. These chests will dynamically update the number of items they request (the buffer) based on usage, and this shouldn't be global because different productions need different amounts of items.

HV Providers won't need to be configured 99% of the time. These will use your filter + bar + empty slot approach to detect which items should be sent to the network. The buffer size on these items will be dynamic and the limit will be globally shared for that item type. The buffer size will always be rounded up to the nearest stack so users will have to use normal Network Chests if they want to buffer less than a stack of items.

The third type of chest is a Network Trash Chest that simply sends all items into the network and is useful for recipe side products.

With these 3 chest types, It seems like there's a way to meet the needs for almost any recipe without having to manually select items. Additionally, these new chest types can be updated much more efficiently than Network Chests which opens the door to optimizations.

What do you think?

**Comment by year6b7a** [comment]
Also since this is a big feature, does it make sense for me to make a branch dedicated to this work and give you push access? That should hopefully make it easier to reduce merge conflicts and move faster without affecting regular releases.

**Comment by bengardner** [comment]
Just to check my understanding...

> HV Requesters 

That would be like a Logistic Requester chest, correct?
It never removes items from the chest, but only tries to request items based on the configured level.  It does not have a "limit" to prioritize requests.
And if you fill the chest with junk then add a new request... tough luck?
Or would it 'trash" anything that it wasn't requesting?
I'd love the UI to look like a Logistic Requester chest, if possible.
Including the inventory view with all the stack manipulation clicks.

> HV Providers 

This would be an "unconfigured" chest with the locked/unlocked state.
It would respect global limits.

> Network Trash Chest 

I'm assuming this would ignore global limits and just send everything.


That all sounds good.

**Comment by bengardner** [comment]
> Also since this is a big feature, does it make sense for me to make a branch dedicated to this work and give you push access? That should hopefully make it easier to reduce merge conflicts and move faster without affecting regular releases.

Yes, that would make sense.

**Comment by year6b7a** [comment]
Sounds good. I made the branch `high-volume-chests` to start on this, and once it's done it can be merged into main.

Yeah HV Requesters are like Logistic Requester chests except HV Requesters don't have a number since that number will be determined automatically (or defined globally if automatic is too hard).
And yes if the chest is filled with junk then it won't do anything. Maybe we can add logic to auto-trash unrequested types of items but I don't think this case can occur accidentally.

Your descriptions for HV Providers and the trash chests are spot on.

The only extra part I'll add is that I'm hoping these new chest types can be efficient enough that we can update a few of them for the same cost as updating a single Network Chest which should increase factory throughput dramatically. Long-term, I see endgame factory setups as having a mall filled with low-throughput network chests while the rest of the factory uses efficient HV chests.
